sunscreen_runtime/src/debugger/data.rs

```diff
@@ -0,0 +1,236 @@
+use crate::Ciphertext;
+use crate::InnerCiphertext;
+use crate::PrivateKey;
+use crate::SealData;
+use crate::WithContext;
+use petgraph::stable_graph::NodeIndex;
+use petgraph::stable_graph::StableGraph;
+use petgraph::Direction::Incoming;
+use seal_fhe::BfvEncryptionParametersBuilder;
+use seal_fhe::Context;
+use seal_fhe::Decryptor;
+use seal_fhe::Modulus;
+use seal_fhe::SecretKey;
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use sunscreen_compiler_common::Operation;
+use sunscreen_compiler_common::Type;
+use sunscreen_compiler_common::{EdgeInfo, NodeInfo};
+
+#[derive(Clone, Serialize, Deserialize)]
+pub enum DebugNodeType {
+    Bfv(BfvNodeType),
+    Zkp(ZkpNodeType),
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct BfvNodeType {
+    pub value: i64,
+    pub data_type: Type,
+    pub noise_budget: Option<u32>,
+    pub coefficients: Vec<Vec<u64>>,
+    pub multiplicative_depth: u64,
+    pub overflowed: Option<bool>,
+    pub noise_exceeded: Option<bool>,
+}
+
+/**
+ * Gets the multiplicative depth of a node in the compilation graph.
+ */
+pub fn get_mult_depth<O>(
+    graph: &StableGraph<NodeInfo<O>, EdgeInfo>,
+    node: NodeIndex,
+    mut depth: u64,
+) -> u64
+where
+    O: Operation,
+{
+    if graph
+        .node_weight(node)
+        .unwrap()
+        .operation
+        .is_multiplication()
+    {
+        depth += 1;
+    }
+
+    let neighbors = graph.neighbors_directed(node, Incoming);
+    if neighbors.clone().count() == 0 {
+        return depth;
+    }
+    let mut max_depth = 0;
+    for neighbor in neighbors.clone() {
+        let neighbor_depth = get_mult_depth(graph, neighbor, depth);
+        max_depth = max_depth.max(neighbor_depth);
+    }
+
+    max_depth
+}
+
+/**
+ * Checks if any coefficients in a polynomial have overflowed.
+ */
+pub fn overflow_occurred<O>(
+    graph: &StableGraph<NodeInfo<O>, EdgeInfo>,
+    node: NodeIndex,
+    p: u64,
+    pk: &PrivateKey,
+    program_data: &[Option<SealData>],
+) -> bool
+where
+    O: Operation,
+{
+    // Overflow only happens after an arithmetic operation involving ciphertexts
+    // If the current node has more than 1 parent, then it's an arithmetic operation, so it can't overflow
+    // If the current node has no parents, then it's an input node, so it can't overflow
+    let mut incoming_neighbors = graph.neighbors_directed(node, Incoming);
+    if incoming_neighbors.clone().count() != 1 {
+        return false;
+    }
+
+    let mut add_overflow = false;
+    let mut mul_overflow = false;
+
+    // Create operands
+    let mut operands: Vec<Vec<Vec<u64>>> = Vec::new();
+    let parent = incoming_neighbors.next().unwrap();
+    let operand_nodes = graph.neighbors_directed(parent, Incoming);
```

Comment:
So, you only need to check for overflow on mul, add, sub all of which are binary operations, right?

So, I'd expect a match statement the operation matching on these cases and calling out to some helper functions. On _ you can return false.

Note we already have methods somewhere to get the binary operand indices for a binary op node that returns you (left_idx, right_idx). You shouldn't need to call neighbors_directed.